### DISCUSSION TITLE
s3scanner returns `not_exist` for everything on most default conditions

### DISCUSSION THREAD

**Discussion by fededamian** [sa7mon/S3Scanner#325] [General]
### Existing Issues

- [X] I've searched the [current issues](https://github.com/sa7mon/S3Scanner/issues)

### Command

docker run --rm ghcr.io/sa7mon/s3scanner -bucket s3scanner-bucketsize

### Observed behavior (what happened?)

I am trying to run this tool, something I remember succeeded in before, but now, no matter what I do, I get the `not_exist` output. Either something is broken or I must be doing something very dumb but even in that case it may be helpful to have a couple command examples in the readme guaranteed to succeed to test the behavior.

I am using the docker image to make sure the infrastructure is not the culprit. See the output:

```
root@ip:~# docker run --rm ghcr.io/sa7mon/s3scanner -version
v3.0.4

root@ip:~# docker run --rm ghcr.io/sa7mon/s3scanner -bucket s3scanner-bucketsize
level=info msg="not_exist | s3scanner-bucketsize"

root@ip:~# docker run --rm ghcr.io/sa7mon/s3scanner -verbose -bucket s3scanner-bucketsize
level=debug msg="bucket not found" bucket_name=s3scanner-bucketsize method="aws.BucketExists()" region=
level=info msg="not_exist | s3scanner-bucketsize"
```

I know the testing buckets work fine because I can open and get the right response (https://s3scanner-bucketsize.s3.amazonaws.com/).

Am I doing something wrong? Or did something change that made the tool break these last days?

Thanks and apologizes for the inconvenience if this is on me.

### Expected behaviour

Getting the `exists` output.

### Debug output

docker run --rm ghcr.io/sa7mon/s3scanner -verbose -bucket s3scanner-bucketsize

level=debug msg="bucket not found" bucket_name=s3scanner-bucketsize method="aws.BucketExists()" region=
level=info msg="not_exist | s3scanner-bucketsize"

### OS Info

Docker Image maintained by the project

**Comment by sa7mon** [comment]
I was not able to reproduce the issue

```
# docker run --rm ghcr.io/sa7mon/s3scanner -bucket s3scanner-bucketsize
Unable to find image 'ghcr.io/sa7mon/s3scanner:latest' locally
latest: Pulling from sa7mon/s3scanner
9fda8d8052c6: Already exists
fb64270fff90: Pull complete
d6b51273a4f6: Pull complete
Digest: sha256:3fe783f409f3b44b748f8f7a69155b813681f3c2cffaf0ecd4e792c64401253e
Status: Downloaded newer image for ghcr.io/sa7mon/s3scanner:latest
level=info msg="exists    | s3scanner-bucketsize | us-east-1 | AuthUsers: [] | AllUsers: [READ]"
# docker run --rm ghcr.io/sa7mon/s3scanner -version                                                                                      
v3.0.4
```

I suspect the IP address you are running the container from is being blocked by AWS, but that's a complete guess. Is this a VPS? Or an IP that also hosts a Tor relay?

**Reply by fededamian** [reply]
Thanks a lot for your answer sa7mon. I ran the docker command on my laptop and it works fine.
The previous test I did, for which I found the issue, was in an AWS EC2 instance with Ubuntu 24.04, nothing else, so no TOR.
The same behavior occurs both running the go binary directly in the Ubuntu box and running the binary in the docker container.
I don't think it is a connectivity issue, if I `curl` the S3 URL I get the right response.

```
root@ip:~# docker run --rm ghcr.io/sa7mon/s3scanner -bucket s3scanner-bucketsize
level=info msg="not_exist | s3scanner-bucketsize"

root@ip:~# curl https://s3scanner-bucketsize.s3.amazonaws.com
<?xml version="1.0" encoding="UTF-8"?>
<ListBucketResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Name>s3scanner-bucketsize</Name><Prefix></Prefix><Marker></Marker><MaxKeys>1000</MaxKeys><IsTruncated>false</IsTruncated><Contents><Key>test-file.txt</Key><LastModified>2019-09-16T23:37:22.000Z</LastModified><ETag>&quot;cbe923fd5deb54c85287df84fe028710&quot;</ETag><Size>43</Size><StorageClass>STANDARD</StorageClass></Contents></ListBucketResult>root@ip-172-31-9-19:~#

root@ip:~# cat /etc/issue
Ubuntu 24.04 LTS \n \l
```
I am running other recon tools and they seem to work fine, so I don't think there is some sort of IP blocking on the egress point.

Any other ideas on what I can look for?

**Reply by sa7mon** [reply]
I'm not sure what's causing your issue. If you'd like to debug further and see the actual HTTP requests and responses, you can setup the development environment as explained in [this section](https://github.com/sa7mon/S3Scanner/tree/main?tab=readme-ov-file#development) of the README.

**Comment by fededamian** [comment]
It's a very curious behavior. I ran the command in the EC2 server earlier today and got many successful results. After repeating everything exactly the same, I got nothing, all `not_exist`. I used MITM proxy and extracted the `curl` command. Then ran it in the EC2 and in my laptop:

##### AWS EC2 Run
```
curl -I -H 'User-Agent: aws-sdk-go-v2/1.27.0 os/linux lang/go#1.19.13 md/GOOS#linux md/GOARCH#amd64 api/s3#1.54.2' --compressed -H 'Amz-Sdk-Invocation-Id: edc28420-e27d-473b-9892-2fb3d3baf28f' -H 'Amz-Sdk-Request: attempt=1; max=3' -H 'X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855' -X HEAD https://s3scanner-bucketsize.s3.us-east-1.amazonaws.com/

HTTP/1.1 404 Not Found
x-amz-request-id: R22WEAJX5WPA82RP
x-amz-id-2: MdATZv1FO5HCgs+OSRz/wV1nad+HKedtrSfK051X6QfWv4wTyHugF7j4VBEyJKxgcqEBmaPgBb4=
Content-Type: application/xml
Date: Thu, 13 Jun 2024 18:13:15 GMT
Server: AmazonS3
```

##### Laptop Run
```
curl -I -H 'User-Agent: aws-sdk-go-v2/1.27.0 os/linux lang/go#1.19.13 md/GOOS#linux md/GOARCH#amd64 api/s3#1.54.2' --compressed -H 'Amz-Sdk-Invocation-Id: edc28420-e27d-473b-9892-2fb3d3baf28f' -H 'Amz-Sdk-Request: attempt=1; max=3' -H 'X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855' -X HEAD https://s3scanner-bucketsize.s3.us-east-1.amazonaws.com/

HTTP/1.1 200 OK
x-amz-id-2: OzEzx0vs7yUpjXV4tFK6B4yqYwvTGKqYFlUU4YBxogFh+/wjCdPFOf2+8jBB4CjtpCc3ou4GiWI=
x-amz-request-id: VM9MN0992V3ZHXPT
Date: Thu, 13 Jun 2024 18:12:31 GMT
x-amz-bucket-region: us-east-1
x-amz-access-point-alias: false
Content-Type: application/xml
Server: AmazonS3
```
The result from the EC2 machine is receiving a `404 Not found`, but if I run the exact same curl from my laptop and home connection I get a `200 OK`.

Do you think AWS may be detecting this in some way and be blocking it with the `Not Found`?

**Reply by sa7mon** [reply]
How strange. I just ran your curl command from my home IP and it worked fine (200 OK). I'm really not sure why AWS is behaving this way, but it's out of my hands at this point.

**Comment by m4ra7h0n** [comment]
Same problem.

**Reply by sa7mon** [reply]
Since this discussion began, I have also observed this behavior with AWS specifically. What I believe is happening is that AWS will not send a 429 if you have checked too many buckets in a given timeframe. It will just start saying that all buckets do not exist. 

The only remedy for this that I have found is to switch to another IP address. Let me know if you find a way around this.

**Reply by defensivedepth** [reply]
@sa7mon So it is not clear what the limits are? I am seeing this too, and am trying to find a way to stay below the limit. Also, I wonder if this is an issue if using AWS creds. Haven't tried that yet.

*edit - doing a straight curl request confirms the existence. I wonder if it is only doing the obfuscation if using the SDK.

**Reply by sa7mon** [reply]
@defensivedepth I have not seen any published limits, but perhaps the docs provide that information somewhere. I have seen in the docs previously (but of course can't find references now) that any rate limits in place are increased if you are signing requests (using credentials).

It is frustrating that they don't just return a 429, especially when using the SDK which is supposed to handle rate limiting automatically. Perhaps the automatic rate limiting does work but only if you are using a single client? Not sure.

**Reply by sa7mon** [reply]
I'm working on an overhaul of the S3Scanner docs currently and I'll be sure to add a section about this. It would be good to do an experiment at some point to see if we can track down exactly where the limits lie.

**Reply by defensivedepth** [reply]
The system/IP I was working with yesterday is still returning `not_found`,  16 hours later.  So it's not just a short hourly block. 

I was doing some testing yesterday on checking for existence via curl, then using the SDK for the rest, but wasn't very successful, yet.

**Reply by sa7mon** [reply]
Very interesting. Here are some factors I think go into the decision to block:
- authenticated vs anonymous requests
- rate of API requests
- percentage of "bad" responses - i.e. bucket doesn't exist vs listing bucket contents
- IP address - "soft block" may only affect a single source IP even when using the same credentials from multiple locations

This is all just speculation so far.

**Reply by defensivedepth** [reply]
Have you seen authenticated requests being blocked eventually?

**Comment by sa7mon** [comment]
I have not, but I also haven't tested it very thoroughly.

**Comment by defensivedepth** [comment]
@sa7mon I did some more testing today with auth requests and was blocked very quickly.  I'm not sure `aws.BucketExists()` is a viable strategy at this point.

**Comment by sa7mon** [comment]
Do you have any idea how many requests were sent and how quickly?

**Reply by defensivedepth** [reply]
bucket list was around 350 entries, default threads of 4.

Trying this morning, and bumping the threads down to 1, I seem to be seeing better results.

**Reply by defensivedepth** [reply]
eh middling results again. I'm starting to look at just simply doing GET requests. Alot less efficient, but seems to be more reliable.

**Reply by sa7mon** [reply]
Thanks for the rough info. One of my long-term tasks is to try and determine exactly where the limits lie using some of the test buckets as known-good ones, checking those periodically while scanning a list.
